src/main/java/com/bobocode/blyznytsia/bibernate/query/MethodNameToQueryParser.java

```diff
@@ -0,0 +1,95 @@
+package com.bobocode.blyznytsia.bibernate.query;
+
+import static com.bobocode.blyznytsia.bibernate.util.EntityUtil.resolveEntityTableName;
+
+import com.bobocode.blyznytsia.bibernate.exception.RepositoryGenerationException;
+import com.bobocode.blyznytsia.bibernate.util.EntityUtil;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * A helper class that generates native SQL queries by parsing method names
+ * that follow the Spring Data JPA naming convention
+ */
+@Slf4j
+public class MethodNameToQueryParser {
+  private static final int WHERE_CLAUSE_POS = 0;
+  private static final int ORDER_BY_CLAUSE_POS = 1;
+  Class<?> entityType;
+
+  public MethodNameToQueryParser(Class<?> entityType) {
+    this.entityType = entityType;
+  }
+
+  public String generateQueryByName(String methodName) {
+    var strWithoutFindBy = methodName.substring(9);
+    var tokens = strWithoutFindBy.split("OrderBy");
+    var queryBuilder = new QueryBuilder(resolveEntityTableName(entityType));
+    parseWhereTokens(tokens[WHERE_CLAUSE_POS], queryBuilder);
+    if (tokens.length > 1) {
+      parseOrderByTokens(tokens[ORDER_BY_CLAUSE_POS], queryBuilder);
+    }
+    return queryBuilder.buildNativeSqlQuery();
+  }
+
+  private void parseOrderByTokens(String orderClauseTokens, QueryBuilder queryBuilder) {
+    if (orderClauseTokens.isBlank()) {
+      return;
+    }
+    queryBuilder.orderBy(resolveColumnName(orderClauseTokens, "Desc"));
+    if (orderClauseTokens.endsWith("Desc")) {
+      queryBuilder.descending();
+    }
+  }
+
+  private void parseWhereTokens(String whereClauseTokens, QueryBuilder builder) {
+    List<String> tokens = new ArrayList<>();
+    var matcher = Pattern.compile("(?<=And|Or)|(?=And|Or|$)").matcher(whereClauseTokens);
+    var start = 0;
+    while (matcher.find()) {
+      int end = matcher.start();
+      if (end != 0) {
+        tokens.add(whereClauseTokens.substring(start, end));
+        start = end;
+      }
+    }
+    tokens.forEach(token -> parseSingleWhereToken(token, builder));
+  }
+
+  private void parseSingleWhereToken(String token, QueryBuilder builder) {
+    log.trace("processing token : " + token);
+    if (token == null || token.isBlank()) {
+      throw new RepositoryGenerationException("invalid token value");
+    } else if (token.equals("And")) {
```

Comment:
can we replace these bunch of else if with swich with pattern maching?

Reply:
for example:

```
private void parseSingleWhereToken(String token, QueryBuilder builder) {
    if (token == null || token.isBlank()) {
      throw new RepositoryGenerationException("invalid token value");
    }
   log.trace("processing token : " + token);
    switch (token) {
      case "And" -> builder.and();
      case "Or" -> builder.or();
      case  String t && isNotNullToken(token)-> builder.isNotNull(resolveColumnName(token, "(IsNotNull|NotNull)"));
      case  String t && isNullToken(token) -> builder.isNull(resolveColumnName(token, "(IsNull|Null)"));
      default -> builder.isEqual(resolveColumnName(token, "(Is|Equals)"));
    }
  }

  private boolean isNotNullToken(String token) {
    return token.endsWith("IsNotNull") || token.endsWith("NotNull");
  }

  private boolean isNullToken(String token) {
    return token.endsWith("IsNull") || token.endsWith("Null");
  }
```

Reply:
Unfortunately, pattern matching is a preview feature, and we've disabled it due to conflicts with other maven dependencies. I've refactored the method to be a bit more readable, hope it'll help